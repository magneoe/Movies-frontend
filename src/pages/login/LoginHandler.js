import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import {login} from './actions';


export class LoginHandler extends Component {
    constructor(props){
        super(props);
        this.state = {
            message: '',
        }
    }
    onSubmit = (event) => {
        event.preventDefault();
        let formData = new FormData();
        formData.append('email', document.getElementById("email").value);
        formData.append('password', document.getElementById("password").value);
        //validate
        if(this.isInputValid(formData)){
             //Attempt login
            this.props.actions.login(formData);
        }
        else
            this.setState({message: 'One or more fields are empty!'});
       
    }
    isInputValid = (formData) => {
        const email = formData.get('email');
        const password = formData.get('password');
        if(!email || email.split('').findIndex((e) => e === '@') === -1)
            return false;
        if(!password || password.length < 1)
            return false;
        return true;
    } 
    render() {
        return (
            <div>
                <h3>Login with Username and Password</h3>
                <h4>{this.state.message}</h4>
                <form method="POST" name="loginFormData" id="loginFormData">
                    <table>
                        <tbody>
                            <tr>
                                <td>User:</td>
                                <td>
                                    <input type='text' name='email' id="email" />
                                </td>
                            </tr>
                            <tr>
                                <td>Password:</td>
                                <td>
                                    <input type='password' name='password' id="password" />
                                </td>
                            </tr>
                            <tr>
                                <td colSpan='2'>
                                    <input name="submit" id="loginSubmit" type="submit" value="Login" onClick={this.onSubmit} />
                                </td>
                            </tr>
                        </tbody>
                    </table>
                </form>
            </div>
        );
    }
}

export default connect(
    (state) => ({
    }),
    (dispatch) => ({
        actions: bindActionCreators(Object.assign({},
            { login }), dispatch)

    })
)(LoginHandler)