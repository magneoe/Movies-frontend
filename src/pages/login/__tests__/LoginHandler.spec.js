import React from 'react';

import Enzyme, { shallow, mount } from 'enzyme';
import ShallowRenderer from 'react-test-renderer/shallow';
import Adapter from 'enzyme-adapter-react-16';
import sinon from 'sinon';
import { LoginHandler } from '../LoginHandler';

beforeAll(() => {
    Enzyme.configure({ adapter: new Adapter() });
});
beforeEach(() => {

});

describe("LoginHandler component", () => {

    describe("The Display unit", () => {
        it("Should render successfully", () => {
            const wrapper = shallow(<LoginHandler />);
            expect(wrapper).toMatchSnapshot();
        });
    });
    describe("IsInputValid function", () => {
        it("Should be valid input", () => {
            let formData = new FormData();
            formData.append('email', 'test@testmail.com');
            formData.append('password', 'testpassword');

            const wrapper = shallow(<LoginHandler />);
            expect(wrapper.instance().isInputValid(formData)).toBe(true);
        });
        it("Should be invalid input", () => {
            //No email
            let formData = new FormData();
            formData.append('password', 'testpassword');

            const wrapper = shallow(<LoginHandler />);
            expect(wrapper.instance().isInputValid(formData)).toBe(false);

            //No email set
            formData = new FormData();
            formData.append('email', '');
            formData.append('password', 'testpassword');

            expect(wrapper.instance().isInputValid(formData)).toBe(false);

            //No @ in email
            formData = new FormData();
            formData.append('email', 'test.com');
            formData.append('password', 'testpassword');

            expect(wrapper.instance().isInputValid(formData)).toBe(false);

            //No password set
            formData = new FormData();
            formData.append('email', 'test@test.com');
            formData.append('password', '');

            expect(wrapper.instance().isInputValid(formData)).toBe(false);
        });
    });

    describe("Simulate login submit", () => {
        it("Should successfully dispatch the action on valid inputs", () => {
            const spy = sinon.spy();
            const wrapper = mount(<LoginHandler actions={{login: spy}}/>, { attachTo: document.body });

            wrapper.find('#email').instance().value = "Test@test.com";
            wrapper.find('#password').instance().value = "Test";
            
            wrapper.find('#loginSubmit').simulate('click',  
                { preventDefault() {} } );

            expect(spy.callCount).toBe(1);

        });

        it("Should not dispatch the action on invalid inputs", () => {
            //No password set
            const spy = sinon.spy();
            const wrapper = mount(<LoginHandler actions={{login: spy}}/>, { attachTo: document.body });
            wrapper.find('#email').value = "Test";
            wrapper.find('#loginSubmit').simulate('click', 
                { preventDefault() {} });

            expect(spy.callCount).toBe(0);
        });
    })

});


