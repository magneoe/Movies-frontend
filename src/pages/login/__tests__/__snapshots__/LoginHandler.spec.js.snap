// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginHandler component The Display unit Should render successfully 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LoginHandler />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h3>
          Login with Username and Password
        </h3>,
        <h4>
          
        </h4>,
        <form
          id="loginFormData"
          method="POST"
          name="loginFormData"
        >
          <table>
            <tbody>
              <tr>
                <td>
                  User:
                </td>
                <td>
                  <input
                    id="email"
                    name="email"
                    type="text"
                  />
                </td>
              </tr>
              <tr>
                <td>
                  Password:
                </td>
                <td>
                  <input
                    id="password"
                    name="password"
                    type="password"
                  />
                </td>
              </tr>
              <tr>
                <td
                  colSpan="2"
                >
                  <input
                    id="loginSubmit"
                    name="submit"
                    onClick={[Function]}
                    type="submit"
                    value="Login"
                  />
                </td>
              </tr>
            </tbody>
          </table>
        </form>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Login with Username and Password",
        },
        "ref": null,
        "rendered": "Login with Username and Password",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "",
        },
        "ref": null,
        "rendered": "",
        "type": "h4",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <table>
            <tbody>
              <tr>
                <td>
                  User:
                </td>
                <td>
                  <input
                    id="email"
                    name="email"
                    type="text"
                  />
                </td>
              </tr>
              <tr>
                <td>
                  Password:
                </td>
                <td>
                  <input
                    id="password"
                    name="password"
                    type="password"
                  />
                </td>
              </tr>
              <tr>
                <td
                  colSpan="2"
                >
                  <input
                    id="loginSubmit"
                    name="submit"
                    onClick={[Function]}
                    type="submit"
                    value="Login"
                  />
                </td>
              </tr>
            </tbody>
          </table>,
          "id": "loginFormData",
          "method": "POST",
          "name": "loginFormData",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <tbody>
              <tr>
                <td>
                  User:
                </td>
                <td>
                  <input
                    id="email"
                    name="email"
                    type="text"
                  />
                </td>
              </tr>
              <tr>
                <td>
                  Password:
                </td>
                <td>
                  <input
                    id="password"
                    name="password"
                    type="password"
                  />
                </td>
              </tr>
              <tr>
                <td
                  colSpan="2"
                >
                  <input
                    id="loginSubmit"
                    name="submit"
                    onClick={[Function]}
                    type="submit"
                    value="Login"
                  />
                </td>
              </tr>
            </tbody>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <tr>
                  <td>
                    User:
                  </td>
                  <td>
                    <input
                      id="email"
                      name="email"
                      type="text"
                    />
                  </td>
                </tr>,
                <tr>
                  <td>
                    Password:
                  </td>
                  <td>
                    <input
                      id="password"
                      name="password"
                      type="password"
                    />
                  </td>
                </tr>,
                <tr>
                  <td
                    colSpan="2"
                  >
                    <input
                      id="loginSubmit"
                      name="submit"
                      onClick={[Function]}
                      type="submit"
                      value="Login"
                    />
                  </td>
                </tr>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <td>
                      User:
                    </td>,
                    <td>
                      <input
                        id="email"
                        name="email"
                        type="text"
                      />
                    </td>,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "User:",
                    },
                    "ref": null,
                    "rendered": "User:",
                    "type": "td",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": <input
                        id="email"
                        name="email"
                        type="text"
                      />,
                    },
                    "ref": null,
                    "rendered": Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "id": "email",
                        "name": "email",
                        "type": "text",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": "input",
                    },
                    "type": "td",
                  },
                ],
                "type": "tr",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <td>
                      Password:
                    </td>,
                    <td>
                      <input
                        id="password"
                        name="password"
                        type="password"
                      />
                    </td>,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Password:",
                    },
                    "ref": null,
                    "rendered": "Password:",
                    "type": "td",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": <input
                        id="password"
                        name="password"
                        type="password"
                      />,
                    },
                    "ref": null,
                    "rendered": Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "id": "password",
                        "name": "password",
                        "type": "password",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": "input",
                    },
                    "type": "td",
                  },
                ],
                "type": "tr",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <td
                    colSpan="2"
                  >
                    <input
                      id="loginSubmit"
                      name="submit"
                      onClick={[Function]}
                      type="submit"
                      value="Login"
                    />
                  </td>,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <input
                      id="loginSubmit"
                      name="submit"
                      onClick={[Function]}
                      type="submit"
                      value="Login"
                    />,
                    "colSpan": "2",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "id": "loginSubmit",
                      "name": "submit",
                      "onClick": [Function],
                      "type": "submit",
                      "value": "Login",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "input",
                  },
                  "type": "td",
                },
                "type": "tr",
              },
            ],
            "type": "tbody",
          },
          "type": "table",
        },
        "type": "form",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h3>
            Login with Username and Password
          </h3>,
          <h4>
            
          </h4>,
          <form
            id="loginFormData"
            method="POST"
            name="loginFormData"
          >
            <table>
              <tbody>
                <tr>
                  <td>
                    User:
                  </td>
                  <td>
                    <input
                      id="email"
                      name="email"
                      type="text"
                    />
                  </td>
                </tr>
                <tr>
                  <td>
                    Password:
                  </td>
                  <td>
                    <input
                      id="password"
                      name="password"
                      type="password"
                    />
                  </td>
                </tr>
                <tr>
                  <td
                    colSpan="2"
                  >
                    <input
                      id="loginSubmit"
                      name="submit"
                      onClick={[Function]}
                      type="submit"
                      value="Login"
                    />
                  </td>
                </tr>
              </tbody>
            </table>
          </form>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Login with Username and Password",
          },
          "ref": null,
          "rendered": "Login with Username and Password",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "",
          },
          "ref": null,
          "rendered": "",
          "type": "h4",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <table>
              <tbody>
                <tr>
                  <td>
                    User:
                  </td>
                  <td>
                    <input
                      id="email"
                      name="email"
                      type="text"
                    />
                  </td>
                </tr>
                <tr>
                  <td>
                    Password:
                  </td>
                  <td>
                    <input
                      id="password"
                      name="password"
                      type="password"
                    />
                  </td>
                </tr>
                <tr>
                  <td
                    colSpan="2"
                  >
                    <input
                      id="loginSubmit"
                      name="submit"
                      onClick={[Function]}
                      type="submit"
                      value="Login"
                    />
                  </td>
                </tr>
              </tbody>
            </table>,
            "id": "loginFormData",
            "method": "POST",
            "name": "loginFormData",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <tbody>
                <tr>
                  <td>
                    User:
                  </td>
                  <td>
                    <input
                      id="email"
                      name="email"
                      type="text"
                    />
                  </td>
                </tr>
                <tr>
                  <td>
                    Password:
                  </td>
                  <td>
                    <input
                      id="password"
                      name="password"
                      type="password"
                    />
                  </td>
                </tr>
                <tr>
                  <td
                    colSpan="2"
                  >
                    <input
                      id="loginSubmit"
                      name="submit"
                      onClick={[Function]}
                      type="submit"
                      value="Login"
                    />
                  </td>
                </tr>
              </tbody>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <tr>
                    <td>
                      User:
                    </td>
                    <td>
                      <input
                        id="email"
                        name="email"
                        type="text"
                      />
                    </td>
                  </tr>,
                  <tr>
                    <td>
                      Password:
                    </td>
                    <td>
                      <input
                        id="password"
                        name="password"
                        type="password"
                      />
                    </td>
                  </tr>,
                  <tr>
                    <td
                      colSpan="2"
                    >
                      <input
                        id="loginSubmit"
                        name="submit"
                        onClick={[Function]}
                        type="submit"
                        value="Login"
                      />
                    </td>
                  </tr>,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      <td>
                        User:
                      </td>,
                      <td>
                        <input
                          id="email"
                          name="email"
                          type="text"
                        />
                      </td>,
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "User:",
                      },
                      "ref": null,
                      "rendered": "User:",
                      "type": "td",
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": <input
                          id="email"
                          name="email"
                          type="text"
                        />,
                      },
                      "ref": null,
                      "rendered": Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "host",
                        "props": Object {
                          "id": "email",
                          "name": "email",
                          "type": "text",
                        },
                        "ref": null,
                        "rendered": null,
                        "type": "input",
                      },
                      "type": "td",
                    },
                  ],
                  "type": "tr",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      <td>
                        Password:
                      </td>,
                      <td>
                        <input
                          id="password"
                          name="password"
                          type="password"
                        />
                      </td>,
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "Password:",
                      },
                      "ref": null,
                      "rendered": "Password:",
                      "type": "td",
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": <input
                          id="password"
                          name="password"
                          type="password"
                        />,
                      },
                      "ref": null,
                      "rendered": Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "host",
                        "props": Object {
                          "id": "password",
                          "name": "password",
                          "type": "password",
                        },
                        "ref": null,
                        "rendered": null,
                        "type": "input",
                      },
                      "type": "td",
                    },
                  ],
                  "type": "tr",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <td
                      colSpan="2"
                    >
                      <input
                        id="loginSubmit"
                        name="submit"
                        onClick={[Function]}
                        type="submit"
                        value="Login"
                      />
                    </td>,
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": <input
                        id="loginSubmit"
                        name="submit"
                        onClick={[Function]}
                        type="submit"
                        value="Login"
                      />,
                      "colSpan": "2",
                    },
                    "ref": null,
                    "rendered": Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "id": "loginSubmit",
                        "name": "submit",
                        "onClick": [Function],
                        "type": "submit",
                        "value": "Login",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": "input",
                    },
                    "type": "td",
                  },
                  "type": "tr",
                },
              ],
              "type": "tbody",
            },
            "type": "table",
          },
          "type": "form",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
